### ISSUE TITLE
go: don't change the libraries in 1.18

### ISSUE THREAD

**Issue by robpike** [golang/go#48918]
The 1.18 release of the Go language is likely to include by far the biggest change to the language since its creation: parametric polymorphism, colloquially called generics. There has been much discussion about how the core libraries will adapt, and how to make that adaptation. See #45955 and #48594 for example, and there are others already and sure to be more soon.

How to use these ideas in the standard library requires great thought and planning. Putting them in the library now also adds a significant burden to rolling out the release.

I propose that we do **not** update the libraries in 1.18.

The reason is simple and compelling: It's too much to do all at once, and we might get it wrong. The language changes have been worked on in some form for over a decade, but the library changes are very new, and we have no experience with the use of the new types in Go on which to base a strong case for their design. Yes, we can reason about them at length and much has been done. Experience with other languages helps, but one thing Go has taught us is that it grows its own ways of doing things.

For generics, we don't know what those new ways are yet. Also, the compatibility promise makes the cost of getting any detail wrong quite high. We should wait, watch, and learn.

Instead, I propose we still design, build, test, and use new libraries for slices, maps, channels, and so on, but start by putting them in the golang/x/exp repository. That way, these new libraries - which are truly experimental at this stage - can be tested in production, but can be changed, adapted, and grown for a cycle or two, letting the whole community try them out, if they are interested and willing to accept a little instability, without requiring every detail of every component to be ready from day one. Once they have soaked a bit, and updated through experience, we move them into the main repo as we have done with other externally-grown packages, but with the confidence that they work well in practice and are deserving of our compatibility promise.

I realize everyone wants to get their hands on the fun of the new language feature, and is looking forward to fixing some of the issues in the core libraries that will be less clumsy once it arrives, but I strongly believe it is best to take it slow for now. Use, learn, study, and move cautiously.

**Comment by ALTree** [comment]
Agree. For example, it looked like we were going to add a few new things like `constraints.Slice` and `constraints.Map`, which I found really ugly (but thought were necessary), but now after #48424 it looks like they may be not needed (See https://github.com/golang/go/issues/48424#issuecomment-937158561), which is great! And this happened last week (#48424 was accepted 5 days ago). 

To me this suggest we don't yet fully understand how the current generic implementation (and possible refinements that may be implemented in the near future) would influence these new libraries design; so it may be premature to commit to an API for them. Putting them in `x/exp` seems like a prudent move.

**Comment by andig** [comment]
> How to use these ideas in the standard library requires great thought and planning. Putting them in the library now also adds a significant burden to rolling out the release.
>I propose that we do not update the libraries in 1.18.

I understood rob doesn't want to change the _existing_ libraries. It seems a good idea to introduce new ones and potentially NOT promise to keep them or their api for one or two releases. That will gain more real-world feedback.

**Comment by thepudds** [comment]
> For example, it looked like we were going to add a few new things like `constraints.Slice` and `constraints.Map`, which I found really ugly (but thought were necessary), but now after #48424 it looks like they may be not needed (See https://github.com/golang/go/issues/48424#issuecomment-937158561), which is great! And this happened last week (#48424 was accepted 5 days ago). 

Minor: that proposal is in "likely accept", and not yet accepted, though there is a prototype.

Separately, if the new packages do start at `exp` (or `x/…`), hopefully they could be tagged to clearly communicate to tools and to humans any backwards compatibility intentions that apply at that import path location (e.g., perhaps something like v0.1.0 ~now, and v1.0.0 when go1.18 is released, with an option to move to v2 at the `exp` location if warranted, which helps with library importers and so on).

When those capabilities land in the standard library, or if there is a v1 to v2 change at `exp`, ideally migration for clients could be handled via cmd/go’s hopefully-not-too-distant automatic API migration (initially described [here](https://research.swtch.com/vgo-import#automatic_api_updates)  and more recently discussed [here](https://twitter.com/_rsc/status/1414229736688177156)).

**Comment by neild** [comment]
Putting the proposed Go 1.13 error changes in `golang.org/x/exp/xerrors` was very useful for validating the API before committing to it.

However, applying the compatibility guarantee to the `x/exp` package made it impossible to iterate on the API. The eventual changes to the `errors` package in Go 1.13 differed from the `xerrors` prototype in various ways, but `xerrors` only permitted validating that initial draft. I think this was a mistake.

Testing out new designs in `x/exp` or some other non-std repo is an excellent idea, but next time let's do it in a way that allows us to iterate on the design, whether that means frequent versioning (`set/v1`, `set/v2`, etc.) or accepting that incompatible changes will be made to the experimental package to keep it in sync with the eventual design.

**Comment by hherman1** [comment]
Why was this closed?

**Comment by robpike** [comment]
Yes, it was a mistake, sorry. I was trying to close a different issue and got toomanywindowsbefuddled.

**Comment by mvdan** [comment]
> allows us to iterate on the design, whether that means frequent versioning (`set/v1`, `set/v2`, etc.) or accepting that incompatible changes will be made

I think backwards compatibility is important if we want larger projects to test these APIs properly. Especially for stable libraries, adding dependencies to unstable packages is dangerous, because your library can break at any point via MVS updates.

I was initially envisioning a module like `x/generics` being bumped to `x/generics/v2` if needed, but I think per-package modules like `x/set/v2` might work even better - then you can major-version them separately without having to bump everything at once.

**Comment by changkun** [comment]
> Go was public for three years before 1 was released and available internally at google for longer than that.

And? It is a weak argument against Go 1.18 alpha and beta versions may even longer than that.

**Comment by DmitriyMV** [comment]
I think this worked well for `golang.org/x/xerrors`. With modules it can work even better.

**Comment by bitfield** [comment]
> I agree in general [that real-world use often generates improvements to a design], but in the case of slices, the package contains a lot of functions that I and others have already written piecemeal for particular uses.

Dijkstra reminds us (in an essay aptly called [The Humble Programmer](https://www.cs.utexas.edu/users/EWD/transcriptions/EWD03xx/EWD340.html)):
> The competent programmer is fully aware of the strictly limited size of his own skull; therefore he approaches the programming task in full humility.

In other words, however smart and capable one or two individual Go programmers might be, they can't necessarily anticipate what kind of programs a million Gophers will want to write, or how they'll write them. Even the beautiful, pragmatic designs of the standard library are occasionally found to be wanting in some minor respect. I think Dijkstra would have approved of Go, because a kind of [humility](https://bitfieldconsulting.com/golang/tao-of-go) is built into it. We all need reminding of the value of humility once in a while (he said, humbly referencing his own blog post).

**Comment by Sivyer9303** [comment]
Agree.It's better to put them into exp package before it's ripe.

**Comment by frederikhors** [comment]
I agree. Please proceed with caution, Go is great, **let's go slow but far**!

**Comment by LeGamerDc** [comment]
i agree with this statement, mainly because most job in my mind should be completed with return type inference(which is not supported in 1.18 generics design), and if we could not use generics to complete this job, why do we need it so soon.
